src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -590,13 +572,138 @@ BoundExpression convertArrayElement(BoundExpression element, TypeSymbol elementT
+                builder.Add(result);
+            }
+            return new BoundCollectionInitializerCollectionLiteralExpression(
+                syntax,
+                collectionCreation,
+                naturalTypeOpt: null, // PROTOTYPE: Support natural type.
+                wasTargetTyped: true, // PROTOTYPE: Support natural type.
+                implicitReceiver,
+                builder.ToImmutableAndFree(),
+                targetType)
+            { WasCompilerGenerated = wasCompilerGenerated };
+        }
+
+        private BoundExpression? ConvertCollectionLiteralExpressionToNaturalType(
+            BoundUnconvertedCollectionLiteralExpression node,
+            BindingDiagnosticBag diagnostics,
+            bool returnNullIfNoCommonType) // PROTOTYPE: Remove.
+        {
+            bool hasErrors = node.HasErrors;
+            var syntax = (CollectionCreationExpressionSyntax)node.Syntax;
+            var location = syntax.OpenBracketToken.GetLocation();
+
+            var useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+            // PROTOTYPE: Confirm with LDM that we use the element expressions rather than
+            // the element types to infer best common type.
+            var commonType = BestTypeInferrer.InferBestType(node.Initializers, Conversions, ref useSiteInfo, out _);
+            // PROTOTYPE: Test.
+            diagnostics.Add(location, useSiteInfo);
+
+            TypeSymbol collectionType;
+            if (commonType is null)
+            {
+                if (returnNullIfNoCommonType)
+                {
+                    return null;
+                }
+
+                diagnostics.Add(ErrorCode.ERR_ImplicitlyTypedCollectionLiteralNoBestType, location);
+                collectionType = CreateErrorType();
+                hasErrors = true;
+            }
+            else
+            {
+                /* PROTOTYPE: From BindImplicitArrayCreationExpression().
+            if ((object)bestType == null || bestType.IsVoidType()) // Dev10 also reports ERR_ImplicitlyTypedArrayNoBestType for void.
+            {
+                Error(diagnostics, ErrorCode.ERR_ImplicitlyTypedArrayNoBestType, node);
+                bestType = CreateErrorType();
+            }
+
+            if (bestType.IsRestrictedType())
+            {
+                // CS0611: Array elements cannot be of type '{0}'
+                Error(diagnostics, ErrorCode.ERR_ArrayElementCantBeRefAny, node, bestType);
+            }
+                 */
+
+                // Verify each element can be converted to the common type.
+                foreach (var initializer in node.Initializers)
+                {
+                    // PROTOTYPE: ...
+                }
+
+                // PROTOTYPE: Test with missing List<T>.
+                collectionType = GetWellKnownType(WellKnownType.System_Collections_Generic_List_T, diagnostics, location).Construct(commonType);
+            }
+
+            // PROTOTYPE: hasErrors is dropped.
```

Comment:
maybe prototype comment about binding to the `int capacity` constructor.

Reply:
we'll probably want LDM sign off on that behavior as well.

Reply:
> maybe prototype comment about binding to the int capacity constructor.

Added.